### ISSUE TITLE
EN-DE translation example

### ISSUE THREAD

**Issue by iniverno** [pytorch/fairseq#360]
I have followed the steps in the translation examples section to replicate "Attention is all you need" and "Scaling NMT" (English to German). In the command lines shown in the README, languages are specified to preprocess the data, but **no source or target languages are explicitly specified for training or testing**, it seems that this leads to training a model to translate **German to English**. This si consistent with the output generated by generate.py (text in english).
Following the steps as they are shown I get a BLEU score >30 (but as explained above, it is doing German->English translation). However, when I specify the languages in both training and testing to be EN->DE, the BLEU score is <26. (Both cases training for 35 epochs, averaging last 10). Can someone give more info on this?
Thanks!

**Comment by myleott** [comment]
Maybe you didn't preprocess the data properly? If you take a closer look at the preprocessing instructions [1], you'll see `preprocess.py --source-lang en --target-lang de`.

[1] https://github.com/pytorch/fairseq/tree/master/examples/translation#replicating-results-from-scaling-neural-machine-translation

**Comment by myleott** [comment]
Re: BLEU score <26 for En-De, what command did you use to train? How many updates did you train for?

**Comment by iniverno** [comment]
I strictly followed the instructions from 
https://github.com/pytorch/fairseq/tree/master/examples/translation#replicating-results-from-scaling-neural-machine-translation

Downloaded the dataset from google and preprocessed it with 
$ python preprocess.py --source-lang en --target-lang de \
  --trainpref $TEXT/train.tok.clean.bpe.32000 \
  --validpref $TEXT/newstest2013.tok.bpe.32000 \
  --testpref $TEXT/newstest2014.tok.bpe.32000 \
  --destdir data-bin/wmt16_en_de_bpe32k \
  --nwordssrc 32768 --nwordstgt 32768 \
  --joined-dictionary


The training command I used was the one in that page:
$ python train.py data-bin/wmt16_en_de_bpe32k \
  --arch transformer_vaswani_wmt_en_de_big --share-all-embeddings \
  --optimizer adam --adam-betas '(0.9, 0.98)' --clip-norm 0.0 \
  --lr-scheduler inverse_sqrt --warmup-init-lr 1e-07 --warmup-updates 4000 \
  --lr 0.0005 --min-lr 1e-09 \
  --dropout 0.3 --weight-decay 0.0 --criterion label_smoothed_cross_entropy --label-smoothing 0.1 \
  --max-tokens 3584 \
  --fp16

That command for training does not specify source and target languages. Later, when I use generate.py on the train model, it is producing a text in english.

191448 updates, 35 epochs (in 8gpus).

**Comment by myleott** [comment]
Default language direction is inferred based on the preprocessing. Can you share the filenames in the directory: `data-bin/wmt16_en_de_bpe32k`?

Also can you share the beginning few output lines of the `train.py` command? In particular the order in which the dictionaries are printed is the source->target order.

I just ran the preprocessing commands you listed above and they gave me the correct direction (i.e., En->De). Can you describe your environment a bit more, e.g., what OS, pytorch version, etc.?

**Comment by iniverno** [comment]
The issue of the incorrect order when the languages are not specified disappears with the code in master.
